### PR TITLE
api: added external client certificate authentication

### LINKED ISSUE #1090: Add support for external client certificate validation (mTLS)
**Describe the solution you'd like**

Enable client certificate validation for external clients ([mTLS](https://en.wikipedia.org/wiki/Mutual_authentication))

Since mTLS is done for a virtual server, not a path, annotations in service manifests can not be used. Annotations in the `Ingress` manifest would be possible but be awkward. So the client validation configuration will only be available in the `IngressRoute` objects.

Example;
```yaml
apiVersion: contour.heptio.com/v1beta1
kind: IngressRoute
metadata:
  name: www
  namespace: default
spec:
  virtualhost:
    fqdn: foo.com
    tls:
      secretName: server-secret
      clientValidation:
        secretName: client-secret
  routes:
    - match: /
      services:
        - name: s1
          port: 80
```

**Anything else you would like to add:**

See also #433 

Relevant envoy documentation; https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto.html?highlight=require_client_certificate

Example;
```yaml
      tls_context:
        require_client_certificate: true
        common_tls_context:
          validation_context:
            verify_certificate_spki: j1fFFSBuF7S/Hzk5xsIq7N7wvVm8xdMpMnt0OUsz5rs=
```

In the example Subject Public Key Information (SPKI) is used to validate the client certificate. This does not have to be secret but is generated outside the cluster without any risk of exposing the private key and can be specified in clear text in the annotation.

The spki is generated as described in the envoy doc with;
```
openssl x509 -in default/cert/kahttp.crt -noout -pubkey \
  | openssl pkey -pubin -outform DER \
  | openssl dgst -sha256 -binary \
  | openssl enc -base64
```

**Environment:**
N/A

### PR DESCRIPTION
**Pull request by tsaarni**
Expose support in HTTPProxy to authenticate external clients (downstream) by
validating their client certificates against trusted CA certificate.

Closes #1090

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

### CODE REVIEW

**Comment by tsaarni** [comment]
I have tested this manually end-to-end and it now seems to work correctly, thanks to #2381 :thumbsup:

I'd be interested to add integration tests when the environment is available!

**Comment by youngnick** [comment]
I updated the title a little bit, because we need to be clear that we're not hitting the other external authentication use case here :).

**Comment by jpeach** [comment]
I'll review this, and intend to land it for the 1.4.0 release. I'm a bit sidetracked figuring out how to write and end-to-end test for it, so the review will be a couple of days away.

**Comment by tsaarni** [comment]
Even the manual test procedure is unfortunately rather complicated :-(

Just in case this could help, I wrote down my manual test cases as step-by-step instructions here https://gist.github.com/tsaarni/4bc8d5e03bc180437a1aa6392430675a.
I'm using linux and that may not work on macos as it is.

**Comment by jpeach** [comment]
> Even the manual test procedure is unfortunately rather complicated :-(
> 
> Just in case this could help, I wrote down my manual test cases as step-by-step instructions here https://gist.github.com/tsaarni/4bc8d5e03bc180437a1aa6392430675a.
> I'm using linux and that may not work on macos as it is.

Thanks @tsaarni, that script is a great reference. [This](https://gist.github.com/jpeach/97ab8f44cca3b3f6c2b884e4b5524796) is my test case so far. It assumes that cert-manager is installed, creates a local CA and makes a HTTPS request using the CA from the server certificate.

kind can inject a local build of Contour, and then I can verify the client auth. I'll take some test cases from your script :)

**Comment by youngnick** [comment]
Happy for you to make the call @jpeach.

**Comment by tsaarni** [comment]
I've now rebased and also re-executed my [manual tests](https://gist.github.com/tsaarni/4bc8d5e03bc180437a1aa6392430675a).

File: apis/projectcontour/v1/httpproxy.go [hunk 1]
```diff
@@ -120,6 +120,12 @@ type TLS struct {
 	// backing cluster.
 	// +optional
 	Passthrough bool `json:"passthrough,omitempty"`
+	// ClientValidation defines how to verify the client certificate. This setting:
+	// 1. Enables TLS client certificate validation.
+	// 2. Requires clients to present a TLS certificate (i.e. not optional validation).
+	// 3. Specifies how the client certificate will be validated.
+	// +optional
+	ClientValidation *DownstreamValidation `json:"clientValidation,omitempty"`
```
Comment (stevesloka) [inline apis/projectcontour/v1/httpproxy.go:128]: I wonder if there's a better word to describe what "client" is. I think it's probably best, but it gets confusing with upstream/downstream/client, etc.

File: apis/projectcontour/v1/httpproxy.go [hunk 2]
```diff
@@ -379,6 +385,15 @@ type UpstreamValidation struct {
 	SubjectName string `json:"subjectName"`
 }
 
+// DownstreamValidation defines how to verify the client certificate.
+type DownstreamValidation struct {
```
Comment (stevesloka) [inline apis/projectcontour/v1/httpproxy.go:389]: This is just a struct name so no big deal, but do we need to give it an identity? What if we used this to validate connections elsewhere? Maybe a more generic name that could allow for re-use later?

File: internal/dag/builder.go [hunk 3]
```diff
@@ -1094,6 +1104,23 @@ func (b *Builder) lookupUpstreamValidation(uv *projcontour.UpstreamValidation, n
 	}, nil
 }
 
+func (b *Builder) lookupDownstreamValidation(vc *projcontour.DownstreamValidation, namespace string) (*PeerValidationContext, error) {
+	if vc == nil {
```
Comment (stevesloka) [inline internal/dag/builder.go:1108]: This never gets called.

Comment (tsaarni) [reply internal/dag/builder.go]: I removed the dead if branch `if vc == nil...`

File: apis/projectcontour/v1/httpproxy.go (hunk 1 above)
Comment (tsaarni) [reply apis/projectcontour/v1/httpproxy.go:128]: Sometimes we have used "external" to highlight that it is the client outside of the cluster. 
What about something like this:

```diff
-       // ClientValidation defines how to verify the client certificate. This setting:
+       // ClientValidation defines how to verify the client certificate when an external client establishes TLS connection to Envoy. This setting:
```

File: apis/projectcontour/v1/httpproxy.go (hunk 2 above)
Comment (tsaarni) [reply apis/projectcontour/v1/httpproxy.go:389]: Could you explain bit more what you mean by giving an identity?  I think I didn't understand it :)

There was `UpstreamValidation` before and `DownstreamValidation` mirrors that with the exception that `SubjectName` attibute is not present. Otherwise single generic structure would have worked.

Comment (jpeach) [reply apis/projectcontour/v1/httpproxy.go:389]: The way I was thinking about this was that as we add more API surface, we would eventually converge the two struct names into a common `PeerValidation` like the one we use internally.

**Comment by tsaarni** [comment]
> Another quick question, does this apply to TCP connections as well?

I did not realize this before: the code does not make difference between the two and it does work just fine with TCP connections as well. 

Maybe there should be another unittest in `internal/featuretests/downstreamvalidation_test.go` for this case, though it basically exercises the same code paths from perspective of the downstream validation?

For what it is worth, I added tests in my [manual procedure](https://gist.github.com/tsaarni/4bc8d5e03bc180437a1aa6392430675a)
* test case: successful authentication with tcpproxy in TLS termination mode
* test case: unsuccessful authentication with tcpproxy in TLS termination mode

**Comment by tsaarni** [comment]
There is corner case: `passthrough: true` with `clientValidation`

```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: tcpproxy-passthrough-with-client-auth
spec:
  virtualhost:
    fqdn: protected-tcpproxy-passthrough.127-0-0-101.nip.io
    tls:
      passthrough: true
      clientValidation:
        caSecret: client-root-ca
  tcpproxy:
    services:
      - name: httpbin-protected
        port: 80
```

Currently this does not raise error message in `HTTPProxy` status but I think this should be rejected?

I'll see how to fix it.

**Comment by tsaarni** [comment]
Yet another corner case: `permitInsecure: true` with `clientValidation`

```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: https-with-client-auth-permitinsecure
spec:
  virtualhost:
    fqdn: protected-permitinsecure.127-0-0-101.nip.io
    tls:
      secretName: ingress
      clientValidation:
        caSecret: client-root-ca
  routes:
    - services:
        - name: httpbin-protected
          port: 80
      permitInsecure: true
```

I think this should be rejected as well?

**Comment by jpeach** [comment]
> tcpproxy 

I could go either way on this. Since we are proxying, we could argue that the backend service owns the auth. But this can also be used to make additional guarantees, so it is not conflicting.

> permitInsecure

I don't see anything wrong with this. You would end up with certificate auth required on port 443 but not on 80. Seems weird, but not conflicting.

/cc @youngnick @stevesloka

**Comment by tsaarni** [comment]
Regarding tcpproxy conflict:

There is already a check that covers combination `passthrough: true` and `secretName` which results in error  *tcpproxy: missing tls.passthrough or tls.secretName*.  Shouldn't I extend this check also to cover `passthrough: true` and `clientValidation` as well since that is invalid combination?

The current error message is a bit misleading though since the same error handling logic covers cases when both are missing or both are defined.  I guess it could recognize this and give error such as *tcpproxy: TLS passthrough cannot be combined with tls.secretName or tls.clientValidation*

Regarding permitInsecure:

It is of course OK for me too to allow combination of `permitInsecure` and `clientValidation` in single `HTTPProxy`.  The risk of user unintentionally exposing vhost with and without authentication is probably small.

**Comment by youngnick** [comment]
@tsaarni I'm in favour of marking a HTTPProxy invalid if it contains config that doesn't do anything. So, for `clientValidation: true` and passthrough, we should mark it as invalid.

I also agree that the error message should clearly what's happening, `tcpproxy: TLS passthrough cannot be combined with tls.secretName or tls.clientValidation` is great.

**Comment by tsaarni** [comment]
I've now added following:
* mark `HTTPProxy` as invalid if user configures `passthrough: true` together with `clientValidation` and 
* improved error message for case `tcpproxy: TLS passthrough cannot be combined with tls.secretName or tls.clientValidation`
* test case to `builder_test.go` for *"insert httpproxy w/ tcpproxy in tls termination mode w/ downstream verification"*.  This is when Envoy really terminates TLS on behalf of backend i.e. `passthrough: false`

**Comment by jpeach** [comment]
@tsaarni From now, let's please create new issues for any more changes here.

@stevesloka Are your comments addressed? I'll need to do another review pass for the latest changes but I'll wait for your approval before merging.

File: apis/projectcontour/v1/httpproxy.go (hunk 2 above)
Comment (stevesloka) [reply apis/projectcontour/v1/httpproxy.go:389]: Yeah, what I meant was @jpeach mentioned. Calling the struct something more generic so that it could be reused for other similar items later.

Comment (jpeach) [reply apis/projectcontour/v1/httpproxy.go:389]: @stevesloka Were you thinking that we should make that change now? If we do, that would imply to API users that all the fields currently in `UpstreamValidation` would work, or we would have to add extra validation code to display errors if those fields were used in a downstream context.

Comment (stevesloka) [reply apis/projectcontour/v1/httpproxy.go:389]: I never said merge it with UpstreamValidation. Just ship as is. LGTM

Comment (stevesloka) [reply apis/projectcontour/v1/httpproxy.go:389]: @jpeach  ^^

Comment (jpeach) [reply apis/projectcontour/v1/httpproxy.go:389]: 'kk thanks @stevesloka . I'll resolve this conversation and we can revisit. These names don't affect API compatibility, so we can change them later if we like.

File: internal/dag/builder.go [hunk 4]
```diff
@@ -478,6 +478,12 @@ func (b *Builder) computeHTTPProxy(proxy *projcontour.HTTPProxy) {
 		// tls is valid if passthrough == true XOR secretName != ""
 		tlsValid = tls.Passthrough != !isBlank(tls.SecretName)
 
+		// passthrough cannot be combined with secretName or clientValidation
+		if tls.Passthrough && (!isBlank(tls.SecretName) || (tls.ClientValidation != nil)) {
```
Comment (jpeach) [inline internal/dag/builder.go:482]: I think that it is enough here to check the client validation. The secret name is check for TCP proxy, though that check could arguably be tighter.

```Go
if tls.Passthrough && tls.ClientValidation != nil {
        sw.SetInvalid("Spec.VirtualHost.TLS ...")
        return
}
```

File: internal/dag/builder.go [hunk 5]
```diff
@@ -489,6 +495,16 @@ func (b *Builder) computeHTTPProxy(proxy *projcontour.HTTPProxy) {
 			svhost := b.lookupSecureVirtualHost(host)
 			svhost.Secret = sec
 			svhost.MinProtoVersion = annotation.MinProtoVersion(proxy.Spec.VirtualHost.TLS.MinimumProtocolVersion)
+
+			// Fill in DownstreamValidation when external client validation is enabled.
+			if tls.ClientValidation != nil {
+				dv, err := b.lookupDownstreamValidation(tls.ClientValidation, proxy.Namespace)
+				if err != nil {
+					sw.SetInvalid("TLS client validation policy error: %s", err)
```
Comment (jpeach) [inline internal/dag/builder.go:503]: Let's make this error consistent with other TLS errors:

```
sw.SetInvalid("Spec.VirtualHost.TLS client validation is invalid: %s", err)
```

File: apis/projectcontour/v1/httpproxy.go (hunk 1 above)
Comment (jpeach) [reply apis/projectcontour/v1/httpproxy.go:128]: I think the proposed replacement is an improvements 👍

File: internal/dag/builder.go (hunk 4 above)
Comment (tsaarni) [reply internal/dag/builder.go]: The code around this had changed a bit so I adapted this suggestion a bit, but implemented the idea.

File: internal/dag/builder.go [hunk 6]
```diff
@@ -1094,6 +1110,18 @@ func (b *Builder) lookupUpstreamValidation(uv *projcontour.UpstreamValidation, n
 	}, nil
 }
 
+func (b *Builder) lookupDownstreamValidation(vc *projcontour.DownstreamValidation, namespace string) (*PeerValidationContext, error) {
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply internal/dag/builder.go:1134]: Rebased and changed `lookupDownstreamValidation` accordingly

**Comment by tsaarni** [comment]
Sorry I missed that error, Thank You!
